### PR TITLE
Add version information to data locations

### LINKED ISSUE #104: Add version information to data locations
We should be able to identify when we download different versions of the data files.

### PR DESCRIPTION
**Pull request by nikhilwoodruff**
Fixes #104

### CODE REVIEW

**Review by nikhilwoodruff — COMMENTED** [review]
1. Get version number as in this PR
2. Tag HF uploads with that version number
3. Set gcloud metadata with that version number

**Review by nikhilwoodruff — COMMENTED** [review]
<del>Just a note to double check that this script is run *after* versioning updates.</del>

Confirmed that it is run after: the dataset build and upload action is triggered by changes to pyproject.toml.

File: policyengine_uk_data/utils/data_upload.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    # Upload to GCS
+    credentials, project_id = google.auth.default()
+    storage_client = storage.Client(
+        credentials=credentials, project=project_id
+    )
+    bucket = storage_client.bucket(gcs_bucket_name)
+    for file_path in files:
+        file_path = Path(file_path)
+        blob = bucket.blob(file_path.name)
+        blob.metadata = {"version": version}
```
Comment (nikhilwoodruff) [inline policyengine_uk_data/utils/data_upload.py:74]: Remove

**Comment by nikhilwoodruff** [comment]
Also separate the HF and GCP logic into two separate functions

File: .github/workflows/publish_package.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
+name: Publish package
+on:
+  push:
+    branches: [ main ]
+    paths:
+        - pyproject.toml
+
+jobs:
+  Publish:
```
Comment (anth-volk) [inline .github/workflows/publish_package.yaml:9]: __issue, blocking:__ This action is not dependent upon completion of `push.yaml`. I believe this would publish to PyPI before testing the package.

File: .github/workflows/publish_package.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+      - name: Set up Python
+        uses: actions/setup-python@v2
+        with:
+          python-version: '3.11'
+      - name: Publish a git tag
+        run: ".github/publish-git-tag.sh || true"
+      - name: Install package
+        run: make install
+      - name: Build package
+        run: make
```
Comment (anth-volk) [inline .github/workflows/publish_package.yaml:26]: __nit:__ Explicitly define the `make` script to run here

File: .github/workflows/publish_package.yaml [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      - name: Remove .whl files
+        run: rm dist/*.whl
+      - name: Publish a Python distribution to PyPI
+        uses: pypa/gh-action-pypi-publish@release/v1
+        with:
+          user: __token__
+          password: ${{ secrets.PYPI }}
```
Comment (anth-volk) [inline .github/workflows/publish_package.yaml:33]: __comment:__ Maybe out of scope, but noting here that this form of token passage is insecure, as documented in https://github.com/PolicyEngine/issues/issues/199.

Guide to securely publishing from GH Actions at https://packaging.python.org/en/latest/guides/publishing-package-distribution-releases-using-github-actions-ci-cd-workflows/

File: .github/workflows/versioning.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+name: Versioning updates
+on:
+  push:
+      branches:
+        - main
+
+      paths:
+        - changelog_entry.yaml
+        - "!pyproject.toml"
```
Comment (anth-volk) [inline .github/workflows/versioning.yaml:11]: __issue, blocking:__ Here, too, this doesn't require completion of the `push.yaml` action, causing versioning to occur even if tests fail.

File: .github/workflows/versioning.yaml [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+        runs-on: ubuntu-latest
+        if: |
+          (!(github.event.head_commit.message == 'Update package version'))
+        steps:
+            - name: Checkout repo
+              uses: actions/checkout@v4
+              with:
+                repository: ${{ github.event.pull_request.head.repo.full_name }}
+                ref: ${{ github.event.pull_request.head.ref }}
+                token: ${{ secrets.POLICYENGINE_GITHUB }}
```
Comment (anth-volk) [inline .github/workflows/versioning.yaml:24]: __suggestion:__ I believe this use of the token opens us up to security concerns; can we use the federated token pattern we use elsewhere?

File: policyengine_uk_data/storage/upload_completed_datasets.py [hunk 7]
```diff
@@ -1,65 +1,25 @@
...
-    blob = "local_authority_weights.h5"
-    blob = bucket.blob(blob)
-    blob.upload_from_filename(STORAGE_FOLDER / "local_authority_weights.h5")
-    print(
-        f"Uploaded local_authority_weights.h5 to GCS bucket policyengine-uk-data-private."
+    upload_data_files(
+        files=dataset_files,
+        hf_repo_name="policyengine-uk-data/datasets",
+        hf_repo_type="dataset",
+        gcs_bucket_name="policyengine-uk-data",
```
Comment (anth-volk) [inline policyengine_uk_data/storage/upload_completed_datasets.py:22]: __question:__ I see here this uses `policyengine-uk-data`, but the `upload_data_files` function's `gcs_bucket_name` defaults to `...-data-private`; Would we be required to call `upload_data_files` for all public-facing files first here, then do it again for `-private` with all + non-public microdata?

File: .github/workflows/versioning.yaml (hunk 5 above)
Comment (anth-volk) [inline .github/workflows/versioning.yaml:11]: __issue:__ This filter may be too broad

If I'm understanding correctly, you're looking to prevent the version updating script from creating an endless loop of `versioning.yaml` Action calls. However, a PR that updates the name of the package or relevant pieces of metadata (including Python version numbering) would prevent this script from running, as well.

File: policyengine_uk_data/utils/data_upload.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+def upload_data_files(
+    files: List[str],
+    gcs_bucket_name: str = "policyengine-uk-data-private",
+    hf_repo_name: str = "policyengine/policyengine-uk-data",
+    hf_repo_type: str = "model",
+    version: str = None,
+):
+    if version is None:
+        version = metadata.version("policyengine-uk-data")
```
Comment (anth-volk) [inline policyengine_uk_data/utils/data_upload.py:19]: __issue, blocking:__ I believe this code raises a couple concerns

This comment applies more broadly across the PR.

If I'm reading this correctly, let's say I make a change to the data and it's approved and pushed and the current version is 1.0.0. The code would do the following:
* Run the `versioning.yaml` script, causing an update to `pyproject.toml` (unless I've modified `pyproject.toml` as part of the PR - I'll return to that case below)
* `versioning.yaml` modifies `pyproject.toml`, causing `push.yaml` and `publish_package.yaml` to run simultaneously; this would cause publication of the package without successful test completion

However, if the code changes ever modify `pyproject.toml` directly, all three scripts would run simultaneously. This would likely cause the code to attempt to write to PyPI at the previous version value, failing, while still also attempting to push without having tested.

The second issue: If for some reason `version` is `None`, I believe this code would attempt to upload files at the old version value, since `importlib` will try to get the version from PyPI.

File: .github/workflows/publish_package.yaml [hunk 9]
```diff
@@ -0,0 +1,35 @@
+name: Publish package
+on:
+  push:
+    branches: [ main ]
+    paths:
```
Comment (anth-volk) [inline .github/workflows/publish_package.yaml:5]: __suggestion:__ In order to create Actions that are dependent upon one another, use the [needs keyword](https://docs.github.com/en/actions/writing-workflows/about-workflows#creating-dependent-jobs) and remove the file dependencies, other than where you prevent `versioning` from looping infinitely

From what I'm seeing, I'd chain them `push` -> `versioning` -> `publish_package`

File: policyengine_uk_data/utils/data_upload.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+        file_path = Path(file_path)
+        if not file_path.exists():
+            raise ValueError(f"File {file_path} does not exist.")
+        hf_operations.append(
+            CommitOperationAdd(
+                path_in_repo=file_path.name,
+                path_or_fileobj=str(file_path),
+            )
+        )
+    commit_info = api.create_commit(
```
Comment (anth-volk) [inline policyengine_uk_data/utils/data_upload.py:57]: __nit:__ Add type hints where feasible, especially for returns like these where the only way to know what's generated is to consult docs

**Review by anth-volk — CHANGES_REQUESTED** [review]
Was great to go through this @nikhilwoodruff. Probably 70% of the comments boil down to: you probably want to use GH Actions's `needs` keyword instead of depending on changes to particular files to avoid strange edge case bugs when sequencing push actions.

File: .github/workflows/publish_package.yaml (hunk 2 above)
Comment (nikhilwoodruff) [reply .github/workflows/publish_package.yaml]: Fixed by moving to after the tests run, within the same action.

File: .github/workflows/publish_package.yaml (hunk 3 above)
Comment (nikhilwoodruff) [reply .github/workflows/publish_package.yaml]: It's actually redundant, removed.

File: .github/workflows/versioning.yaml (hunk 6 above)
Comment (nikhilwoodruff) [reply .github/workflows/versioning.yaml:24]: Filed #107

File: policyengine_uk_data/storage/upload_completed_datasets.py (hunk 7 above)
Comment (nikhilwoodruff) [reply policyengine_uk_data/storage/upload_completed_datasets.py]: Typos here whoops! But the repo/buckets are called:

* policyengine-uk-data-private
* policyengine-us-data

We'll put public data in -data and private data in -data-private

Maybe a better standard in our simplification drive but it works for here and it's what we've been doing so far.

File: policyengine_uk_data/utils/data_upload.py (hunk 8 above)
Comment (nikhilwoodruff) [reply policyengine_uk_data/utils/data_upload.py:19]: This wouldn't happen. Reason:
1. Publish package does **not** upload any data, it just publishes the package to pypi and is so entirely harmless.
2. The versioning action would not run if you change pyproject because the GitHub action explicitly does not run if that file is modified (see the `paths` attribute)

On your second issue: `importlib` will always get the current version of the project, and does not interact with PyPI.

File: policyengine_uk_data/storage/upload_completed_datasets.py (hunk 7 above)
Comment (nikhilwoodruff) [reply policyengine_uk_data/storage/upload_completed_datasets.py]: Few typos here! But we currently have the standard: policyengine-{country}-data for all public data, and -data-private for all non-public data. The UK model publishes private data only.

File: .github/workflows/publish_package.yaml (hunk 9 above)
Comment (nikhilwoodruff) [reply .github/workflows/publish_package.yaml]: This also prevents them being visible on the repo homepage. I think this issue is encapsulated in #108.

File: policyengine_uk_data/utils/data_upload.py (hunk 10 above)
Comment (nikhilwoodruff) [reply policyengine_uk_data/utils/data_upload.py:57]: VS Code infers the type from the method's type hint, it's not just in the docs. Proof below. Our standard is that type hinting is needed iff it's not inferable by a code editor.

image

File: policyengine_uk_data/utils/data_upload.py (hunk 8 above)
Comment (anth-volk) [reply policyengine_uk_data/utils/data_upload.py:19]: Forgot that for this type of repo, publishing to PyPI ultimately isn't the end goal. My mistake on `importlib`, thanks for the correction.

File: .github/workflows/versioning.yaml (hunk 5 above)
Comment (nikhilwoodruff) [reply .github/workflows/versioning.yaml:11]: We can't guarantee tests passing on PR == tests passing on push without making PR tests hours long, and we don't want to do that. I don't think it's concerning, we don't have unexpected side effects as long as the data doesn't upload without passing tests, and that is true here still.

File: .github/workflows/publish_package.yaml (hunk 9 above)
Comment (nikhilwoodruff) [reply .github/workflows/publish_package.yaml]: I meant the homepage of the repo: https://github.com/policyengine/policyengine-uk-data

File: policyengine_uk_data/storage/upload_completed_datasets.py [hunk 11]
```diff
@@ -1,65 +1,25 @@
...
+            raise ValueError(f"File {file_path} does not exist.")
 
-    blob = "local_authority_weights.h5"
-    blob = bucket.blob(blob)
-    blob.upload_from_filename(STORAGE_FOLDER / "local_authority_weights.h5")
-    print(
-        f"Uploaded local_authority_weights.h5 to GCS bucket policyengine-uk-data-private."
+    upload_data_files(
+        files=dataset_files,
+        hf_repo_name="policyengine-uk-data/policyengine-uk-data-private",
```
Comment (anth-volk) [inline policyengine_uk_data/storage/upload_completed_datasets.py:20]: __question:__ Now that we're applying the standard of `policyengine-uk-data` for public files, `..-data-private` for private ones, are the constituency and local area weights private? My understanding was FRS files are private, while the others are public.

**Review by anth-volk — CHANGES_REQUESTED** [review]
One question before approval, in the comments

File: policyengine_uk_data/storage/upload_completed_datasets.py (hunk 11 above)
Comment (nikhilwoodruff) [reply policyengine_uk_data/storage/upload_completed_datasets.py:20]: I'm going to say yes, as they might as well be. They're not explicitly required to be, but they're useless without the private data, and it's some extra complexity to manage two buckets in this repo for now.

Comment (anth-volk) [reply policyengine_uk_data/storage/upload_completed_datasets.py:20]: The challenge we've run into a couple times is this: the more private the underlying datasets become, the harder it is for OPT contributors and/or deves without keys to engage with the code base. Obviously, there's no choice on the FRS, but for other datasets, they don't strictly need to require keyed entry.

I'll propose making them private, as this code does, especially given the high priority of these changes. Down the road, we can determine how to deal with the challenges - whether by just giving people keys, moving other files to a public dataset, something else - as part of https://github.com/PolicyEngine/issues/issues/385 and https://github.com/PolicyEngine/issues/issues/382.

Comment (nikhilwoodruff) [reply policyengine_uk_data/storage/upload_completed_datasets.py:20]: Yes we should make synthetic equivalents- agree with that concern. Though that is out of scope of this PR I think.
